design/cluster-update.md

```diff
@@ -0,0 +1,154 @@
+* Signaling to the operator that the pod has completed its storage updates
+
+We should consider an implementation that allows the OSD pod to refresh it's set of OSDs without restarting the entire pod, but since the replication controller's pod template spec needs to be updated by the operator in order to convey this information to the pod, we may need to live with restarting the pod either way.
+Remember that this will be done one node at a time to mitigate impact to cluster health.
+
+Also, other types of update operations to the cluster (e.g., software upgrade) should be blocked while a cluster update is ongoing.
+
+**TODO:** how does OSD pod signal that it is done with OSD changes?  Perhaps setting a value in a config map?
+
+### Storage Update Process
+This section covers the general sequence for updating storage resources and outlines important considerations for cluster health.
+Before any changes begin, we will temporarily disable scrubbing of placement groups (the process of verifying data integrity of stored objects) to maximize cluster resources that can go to both client I/O and recovery I/O for data migration:
+```console
+ceph osd set noscrub
+ceph osd set nodeep-scrub
+```
+
+Some [Ceph documentation](https://access.redhat.com/documentation/en-us/red_hat_ceph_storage/2/html/administration_guide/adding_and_removing_osd_nodes#recommendations) also recommends limiting backfill and recovery work while storage is being added or removed.
+The intent is to maximize client I/O while sacrificing throughput of data migration.
+I do not believe this is strictly necessary and at this point I would prefer to not limit recovery work in the hopes of finishing data migrations as quickly as possible.
+I suspect that most cluster administrators would not be removing storage when the cluster is under heavy load in the first place.
+This trade-off can be revisited if we see unacceptable performance impact.
+
+#### Adding Storage
+As mentioned previously, we will add one OSD at a time in order to allow the cluster to rebalance itself in a controlled manner and to avoid getting into a situation where there is an unacceptable amount of churn and thrashing.
+Adding a new OSD is fairly simple since the OSD pod logic already supports it:
+* If the entire node is being added, ensure the node is added to the CRUSH map: `ceph osd crush add-bucket {bucket-name} {type}`
+* For each OSD:
+  * Register, format, add OSD to the crush map and start the OSD process like normal
+  * Wait for all placement groups to reach `active+clean` state, meaning data migration is complete.
+
+#### Removing Storage
+Removing storage is a more involved process and it will also be done one OSD at a time to ensure the cluster returns to a clean state.
+Of special note for removing storage is that a check should be performed to ensure that the cluster has enough remaining storage to recover (backfill) the entire set of objects from the OSD that is being removed.
```

Comment:
This seems appropriate to implement as a script in the toolbox. Then the user can explicitly decide which osd(s) they want to remove. Our helper script could automate all the individual commands below. I just don't think we should remove OSDs automatically.

Reply:
As referenced earlier, the operator will orchestrate the automatic removal of OSDs in a safe and controlled manner.